cumulus/pallets/weight-reclaim/src/tests.rs

```diff
@@ -0,0 +1,732 @@
+}
+
+struct TestRecorder {
+	return_values: Box<[usize]>,
+	counter: core::sync::atomic::AtomicUsize,
+}
+
+impl TestRecorder {
+	fn new(values: &[usize]) -> Self {
+		TestRecorder { return_values: values.into(), counter: Default::default() }
+	}
+}
+
+impl sp_trie::ProofSizeProvider for TestRecorder {
+	fn estimate_encoded_size(&self) -> usize {
+		let counter = self.counter.fetch_add(1, core::sync::atomic::Ordering::Relaxed);
+		self.return_values[counter]
+	}
+}
+
+fn setup_test_externalities(proof_values: &[usize]) -> sp_io::TestExternalities {
+	let mut test_ext = new_test_ext();
+	let test_recorder = TestRecorder::new(proof_values);
+	test_ext.register_extension(ProofSizeExt::new(test_recorder));
+	test_ext
+}
+
+#[cfg(feature = "runtime-benchmarks")]
+pub fn setup_test_ext_default() -> sp_io::TestExternalities {
+	setup_test_externalities(&[0; 32])
+}
+
+fn set_current_storage_weight(new_weight: u64) {
+	frame_system::BlockWeight::<Test>::mutate(|current_weight| {
+		current_weight.set(Weight::from_parts(0, new_weight), DispatchClass::Normal);
+	});
+}
+
+fn get_storage_weight() -> Weight {
+	*frame_system::BlockWeight::<Test>::get().get(DispatchClass::Normal)
+}
+
+const CALL: &<Test as frame_system::Config>::RuntimeCall =
+	&RuntimeCall::System(frame_system::Call::set_heap_pages { pages: 0u64 });
+const ALICE_ORIGIN: frame_system::Origin<Test> = frame_system::Origin::<Test>::Signed(99);
+const LEN: usize = 150;
+
+fn new_tx_ext() -> Tx {
+	Tx::new((frame_system::CheckWeight::new(), MockExtensionWithRefund))
+}
+
+fn new_extrinsic() -> generic::CheckedExtrinsic<AccountId, RuntimeCall, Tx> {
+	generic::CheckedExtrinsic {
+		format: generic::ExtrinsicFormat::Signed(99, new_tx_ext()),
+		function: RuntimeCall::System(frame_system::Call::remark { remark: vec![] }),
+	}
+}
+
+#[allow(unused)]
+mod doc {
+	type Runtime = super::Test;
+	use crate::StorageWeightReclaim;
+
+	#[docify::export(Tx)]
+	type Tx = StorageWeightReclaim<
+		Runtime,
+		(
+			frame_system::CheckNonce<Runtime>,
+			frame_system::CheckWeight<Runtime>,
+			// ... all other extensions
+			// No need for `frame_system::WeightReclaim` as the reclaim.
+		),
+	>;
+}
+
+#[test]
+fn basic_refund() {
+	// The real cost will be 100 bytes of storage size
+	let mut test_ext = setup_test_externalities(&[0, 100]);
+
+	test_ext.execute_with(|| {
+		set_current_storage_weight(1000);
+
+		// Benchmarked storage weight: 500
+		let info = DispatchInfo { call_weight: Weight::from_parts(0, 500), ..Default::default() };
+		let mut post_info = PostDispatchInfo::default();
+
+		let tx_ext = new_tx_ext();
+
+		// Check weight should add 500 + 150 (len) to weight.
+		let (pre, _) = tx_ext
+			.validate_and_prepare(ALICE_ORIGIN.clone().into(), CALL, &info, LEN)
+			.unwrap();
+
+		assert_eq!(pre.0, Some(0));
+
+		assert_ok!(Tx::post_dispatch(pre, &info, &mut post_info, LEN, &Ok(())));
+
+		// TODO TODO: assert_eq!(post_info.actual_weight, Weight::from_parts(0, 650));
```

Comment:
Need to fix this and also we probably need to test this with a post dispatch info with a `Some(...)` as `post_info.actual_weight` as well as a `None`, but in practice our `CheckedExtrinsic::apply` + `dispatch_transaction` implementation would guarantee that the weight is always `Some`.

Reply:
Fixed in https://github.com/paritytech/polkadot-sdk/pull/6140/commits/363f7d3cd9ec2e097979cc0ecb1c9541bdcab7e5

The case where `None` is given as `post_info` is also tested.
In such case some unspent is returned. (But ignored later as we can't reduce weight from `None`.)